docs/dev/lwt-implementation.md

```diff
@@ -0,0 +1,166 @@
+```
+
+## Scylla Paxos
+
+Scylla Paxos implementation is more elaborate than the base scheme outlined above, and its inner logic is quite different.
+
+The system.paxos table, besides the standard last promised ballot and last accepted proposal, also stores a most_recent_commit field. This field is used as an opmitization. The learn phase stores the committed value in both the base table and in this field. When the next prepare for this key comes to the node later, this committed value is returned. This allows the coordinator to avoid a separate quorum read in storage_service::cas, provided that the prune process has not yet removed this field. The prune is started when the learn is confirmed from the quorum of nodes. This means that this optimization doesn't apply if LWT requests are rare.
+
+The `storage_proxy::cas` consists of the following main blocks:
+
+1. `begin_and_repair_paxos`
+
+2. quorum_read based on the partition_key and read_command
+
+3. create mutation based on the read data from 2
+
+4. create a proposal based on the ballot from 1 and mutation from 3
+
+5. try to accept the proposal, learn if accepted, sleep/retry if failed
+
+`begin_and_repair_paxos` is similar in purpose to the prepare phase from the Simple Paxos above, but its logic and invariants are very different.
+
+It runs a loop, on each iteration it chooses a new wall-clock based ballot and tries to find a quorum of replicas that satisfy the following three invariant:
+* no replica has promised a greater or equal ballot
+* there were no previously accepted proposals on any replicas OR such an accepted proposal is greater than the most recently committed ballot
+* there are no not-pruned commited proposals OR all replicas agree on its ballot
+
+* It retries with a newer ballot as soon as it gets a first prepare reject, not trying to collect a (still possible) quorum of prepare. This catches the situation when some other proposal with higher ballot is already under way, so we give up early to not compete with it.
+
+The prepare_summary structure:
+	most_recent_commit
+		is populated with the most recent value from the most_recent_commit column from the replicas, participated in the prepare round. If prune finished for this partition key, then this column will not be set on the replicas and in summary.most_recent_commit field.
+
+	most_recent_proposal
+		is populated with the most recent value from the accept_proposal column from the replicas, participated in the prepare round.
+
+## Questions:
+
+* Why we use db::consistency_level::ANY in repair code in begin_and_repair_paxos? Doesn't this mean that we only wait for one replica to confirm? This will be a different replica from the replica where we get the most recent commit in the first place, so in fact we get two confirmations, but it's still not enough for a quorum on a larger replica sets. It's not guaranteed that the quorum read for CAS conditions in storage_proxy::cas would return the latest value.
+
```

Comment:
If some wrote will fail next repair will re-try repairing. No point in waiting.

Reply:
Suppose we have 5 replicas, cl=quorum=3, and only one replica knows the most recent commit -- it's the only replica where learn of the previous paxos round succeeded. We don't do "Finishing incomplete paxos round" in `begin_and_repair_paxos` because `in_progress->ballot.timestamp() <= summary.most_recent_commit->ballot.timestamp())`. We propagate the write to one more replica in `_proxy->mutate_internal(std::move(m), db::consistency_level::ANY, ...)`, all other fail for some reason. Then we proceed to 
`co_await query(schema, cmd, std::move(pr), cl, query_options);` in `storage_proxy::cas` and read stale data from the other 3 replicas.

Next repair will be working with the effects of the new paxos round, with new mutation, this one will never be repaired.

Reply:
Looks like a bug introduced in 61b2e41a23e. The loop inside `begin_and_repair_paxos` should be repeated unconditionally if there was a repair, not only when write fails.